consensus/hotstuff/cruisectl/block_rate_controller_test.go

```diff
@@ -2,25 +2,381 @@ package cruisectl
-func TestStartStop(t *testing.T) {
-	state := mockprotocol.NewState(t)
-	ctl, err := NewBlockRateController(unittest.Logger(), DefaultConfig(), state)
-	require.NoError(t, err)
-
-	ctx, cancel := irrecoverable.NewMockSignalerContextWithCancel(t, context.Background())
-	ctl.Start(ctx)
-	unittest.RequireCloseBefore(t, ctl.Ready(), time.Second, "component did not start")
-	cancel()
-	unittest.RequireCloseBefore(t, ctl.Done(), time.Second, "component did not stop")
+func (bs *BlockRateControllerSuite) TestStartStop() {
+	bs.CreateAndStartController()
+	bs.StopController()
+}
+
+// TestInit_EpochStakingPhase tests initializing the component in the EpochStaking phase.
+// Measurement and epoch info should be initialized, next epoch final view should be nil.
+func (bs *BlockRateControllerSuite) TestInit_EpochStakingPhase() {
+	bs.CreateAndStartController()
+	defer bs.StopController()
+	bs.AssertCorrectInitialization()
+}
+
+// TestInit_EpochStakingPhase tests initializing the component in the EpochSetup phase.
+// Measurement and epoch info should be initialized, next epoch final view should be set.
+func (bs *BlockRateControllerSuite) TestInit_EpochSetupPhase() {
+	nextEpoch := mockprotocol.NewEpoch(bs.T())
+	nextEpoch.On("Counter").Return(bs.epochCounter+1, nil)
+	nextEpoch.On("FinalView").Return(bs.curEpochFinalView+100_000, nil)
+	bs.epochs.Add(nextEpoch)
+
+	bs.CreateAndStartController()
+	defer bs.StopController()
+	bs.AssertCorrectInitialization()
+}
+
+// TestInit_EpochFallbackTriggered tests initializing the component when epoch fallback is triggered.
+// Default proposal delay should be set.
+func (bs *BlockRateControllerSuite) TestInit_EpochFallbackTriggered() {
+	bs.epochFallbackTriggered = true
+	bs.CreateAndStartController()
+	defer bs.StopController()
+	bs.AssertCorrectInitialization()
+}
+
+// TestEpochFallbackTriggered tests epoch fallback:
+//   - the proposal delay should revert to default
+//   - duplicate events should be no-ops
+func (bs *BlockRateControllerSuite) TestEpochFallbackTriggered() {
+	bs.CreateAndStartController()
+	defer bs.StopController()
+
+	// update error so that proposal delay is non-default
+	bs.ctl.lastMeasurement.instErr *= 1.1
+	err := bs.ctl.measureViewRate(bs.initialView+1, time.Now())
+	require.NoError(bs.T(), err)
+	assert.NotEqual(bs.T(), bs.config.DefaultProposalDelay, bs.ctl.ProposalDelay())
+
+	// send the event
+	bs.ctl.EpochEmergencyFallbackTriggered()
+	// async: should revert to default proposal delay
+	require.Eventually(bs.T(), func() bool {
+		return bs.config.DefaultProposalDelay == bs.ctl.ProposalDelay()
+	}, time.Second, time.Millisecond)
+
+	// additional events should be no-ops
+	// (send capacity+1 events to guarantee one is processed)
+	for i := 0; i <= cap(bs.ctl.epochFallbacks); i++ {
+		bs.ctl.EpochEmergencyFallbackTriggered()
+	}
+	assert.Equal(bs.T(), bs.config.DefaultProposalDelay, bs.ctl.ProposalDelay())
```

Comment:
can we expand this test by sending view change events and making sure it didn't affect proposal delay?

Reply:
💡 good idea - will add this